### DISCUSSION TITLE
A proper way to Implement Word-to-ID mapping feature

### DISCUSSION THREAD

**Discussion by nmago** [ueberdosis/tiptap#4793] [General Questions]
Hi everyone!

First and foremost, I would like to express my gratitude to all of you for creating such an impressive text editing framework.

Now, let's dive into the specific challenge I'm facing:

I am currently working on implementing a feature that involves associating unique IDs with each word in my text, IDs map each word to its corresponding translation in DB. While the intention is not to assign IDs to all words initially, the goal is to cover a significant portion in the future.

The approach that crossed my mind is to create invisible nodes wrapping words, with the ID as an attribute. This would look something like:

`I <translation-node id="1234">love</translation-node> programming`

**Question 1:** Is this the optimal approach, or are there alternative methods that might be more effective?

The second aspect of my implementation involves utilizing Tiptap in two modes: read-only mode for regular users and editable mode for administrators. I need to enable users to propose translations for any word by selecting the word (let's assume any text piece), choosing a translation (with the ID provided by the API), and then saving the mapping of this ID for the selected text.

Hence my **2 question is**, how to handle and store proposed mapping from user? For example, if I programmatically wrap selected word into a node in the editor, are there straightforward methods for conflict resolution before saving to my database? Considering that I'm storing content in JSON, would utilizing the collaboration module be a viable solution, or is it more practical to implement merge logic in the backend? Or maybe there is other way to store and then apply changes (with diff?). Keep in mind also, that I would not want to apply proposed changes immediately, but after some kind of moderation.

Your insights and suggestions would help me a lot. Thank you!

**Comment by carsonbird** [comment]
I don't have any helpful solutions but I am trying to do something similar, and your implementation idea of defining each word as a node type is the same path I was going to go down.
